worker/worker-execution/score-worker-execution-impl/src/main/java/io/cloudslang/worker/execution/reflection/ReflectionAdapterImpl.java

```diff
@@ -80,124 +78,121 @@ private static String getExceptionMessage(ControlActionMetadata actionMetadata)
-        return actionMethod;
     }
 
-    private Object[] buildParametersArray(Method actionMethod, Map<String, ?> actionData) {
+    private Object[] buildParametersArray(Method actionMethod, ReadonlyStepActionDataAccessor accessor) {
         String actionFullName = actionMethod.getDeclaringClass().getName() + "." + actionMethod.getName();
         String[] paramNames = cacheParamNames.get(actionFullName);
         if (paramNames == null) {
             paramNames = parameterNameDiscoverer.getParameterNames(actionMethod);
             cacheParamNames.put(actionFullName, paramNames);
         }
-        List<Object> args = new ArrayList<>(paramNames.length);
-        for (String paramName : paramNames) {
+
+        Object[] args = new Object[paramNames.length];
+        for (int counter = 0; counter < args.length; counter++) {
+            String paramName = paramNames[counter];
             if (NON_SERIALIZABLE_EXECUTION_DATA.equals(paramName)) {
-                final Long executionId = getExecutionIdFromActionData(actionData);
-                final Long runningId = getRunningExecutionIdFromActionData(actionData);
-                final Map<String, Object> globalSessionsExecutionData = sessionDataHandler
-                        .getGlobalSessionsExecutionData(executionId);
-                final Map<String, Object> sessionObjectExecutionData = sessionDataHandler
-                        .getSessionsExecutionData(executionId, runningId);
+                final Long executionId = getExecutionIdFromActionData(accessor);
+                final Long runningId = getRunningExecutionIdFromActionData(accessor);
+                final Map<String, Object> globalSessionsExecutionData = sessionDataHandler.getGlobalSessionsExecutionData(executionId);
+                final Map<String, Object> sessionObjectExecutionData = sessionDataHandler.getSessionsExecutionData(executionId, runningId);
 
                 Map<String, Map<String, Object>> nonSerializableExecutionData = new HashMap<>(2);
                 nonSerializableExecutionData.put(GLOBAL_SESSION_OBJECT, globalSessionsExecutionData);
                 nonSerializableExecutionData.put(SESSION_OBJECT, sessionObjectExecutionData);
-
-                args.add(nonSerializableExecutionData);
-
+                args[counter] = nonSerializableExecutionData;
                 // If the control action requires non-serializable session data, we add it to the arguments array
                 // and set the session data as active, so that it won't be cleared
                 sessionDataHandler.setGlobalSessionDataActive(executionId);
                 sessionDataHandler.setSessionDataActive(executionId, runningId);
-                continue;
+            } else {
+                args[counter] = accessor.getValue(paramName);
             }
-            Object param = actionData.get(paramName);
-            args.add(param);
         }
-        return args.toArray(new Object[args.size()]);
+
+        return args;
```

Comment:
That's actually a great improvement.. Good job, Lucian 🥇

Reply:
thanks Flavius 👍